server/src/main/java/org/elasticsearch/action/ActionListener.java

```diff
@@ -290,6 +290,33 @@ public String toString() {
         };
     }
 
+    /**
+     * Adds a wrapper around a listener which catches exceptions thrown by its {@link #onResponse} method and feeds them to its
+     * {@link #onFailure}.
+     */
+    static <Response, BaseResponse extends Response> ActionListener<BaseResponse> wrap(ActionListener<Response> delegate) {
+        return new ActionListener<>() {
+            @Override
+            public void onResponse(BaseResponse response) {
+                try {
+                    delegate.onResponse(response);
+                } catch (Exception e) {
+                    onFailure(e);
+                }
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                delegate.onFailure(e);
+            }
+
+            @Override
+            public String toString() {
+                return "wrapped{" + delegate + "}";
+            }
+        };
+    }
```

Comment:
Should we delegate to the other overloaded implementation of the `wrap` here or it is important to have a different `toString`?

Reply:
it's important to have `delegate.toString()` appear so we can't just `return wrap(delegate::onResponse, delegate::onFailure)`, we'd have to do something like this:

```java
return wrap(new CheckedConsumer<>() {
    @Override
    public void accept(BaseResponse baseResponse) {
        delegate.onResponse(baseResponse);
    }

    @Override
    public String toString() {
        return delegate.toString();
    }
}, new Consumer<>() {
    @Override
    public void accept(Exception exception) {
        delegate.onFailure(exception);
    }

    @Override
    public String toString() {
        return "ibid";
    }
});
```